src/handler/nat_hole_puncher.rs

```diff
@@ -0,0 +1,181 @@
+use crate::{node_info::NodeAddress, packet::MessageNonce, Discv5Error, Enr, IpMode};
+use async_trait::async_trait;
+use delay_map::HashSetDelay;
+use enr::NodeId;
+use futures::{Stream, StreamExt};
+use rand::Rng;
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    net::{IpAddr, SocketAddr, UdpSocket},
+    ops::RangeInclusive,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use thiserror::Error;
+
+/// The expected shortest lifetime in most NAT configurations of a punched hole in seconds.
+pub const DEFAULT_HOLE_PUNCH_LIFETIME: u64 = 20;
+/// The default number of ports to try before concluding that the local node is behind NAT.
+pub const DEFAULT_PORT_BIND_TRIES: usize = 4;
+/// Port range that is not impossible to bind to.
+pub const USER_AND_DYNAMIC_PORTS: RangeInclusive<u16> = 1025..=u16::MAX;
+
+/// An error occurred whilst attempting to hole punch NAT.
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("NAT error, failed as initiator a hole punch attempt, {0}")]
+    Initiator(Discv5Error),
+    #[error("NAT error, failed as relay of a hole punch attempt, {0}")]
+    Relay(Discv5Error),
+    #[error("NAT error, failed as target of a hole punch attempt, {0}")]
+    Target(Discv5Error),
+}
+
+#[async_trait]
+pub trait NatHolePunch {
+    /// A request times out. Should trigger the initiation of a hole punch attempt, given a
+    /// transitive route to the target exists.
+    async fn on_request_time_out(
+        &mut self,
+        relay: NodeAddress,
+        local_enr: Enr, // initiator-enr
+        timed_out_nonce: MessageNonce,
+        target_session_index: NodeAddress,
+    ) -> Result<(), Error>;
+    /// A RelayInit notification is received over discv5 indicating this node is the relay. Should
+    /// trigger sending a RelayMsg to the target.
+    async fn on_relay_init(
+        &mut self,
+        initr: Enr,
+        tgt: NodeId,
+        timed_out_nonce: MessageNonce,
+    ) -> Result<(), Error>;
```

Comment:
An example of target/tgt not being used consistently.

For ``local_enr: Enr, // initiator-enr`` maybe instead we could do ``local_initiator_enr: Enr,`` which would keep people from having to juggle comments of what variables mean in their heads